File: services/core/internal/payment/monobank.go
package payment

import (
	"bytes"
	"context"
	"crypto/ecdsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"
)

const monoAcquiringURL = "https://api.monobank.ua/api/merchant"

// MonobankClient implements Monobank Acquiring
type MonobankClient struct {
	token      string
	httpClient *http.Client
	publicKey  *ecdsa.PublicKey
}

// NewMonobankClient creates Monobank client
func NewMonobankClient(token string) *MonobankClient {
	return &MonobankClient{
		token:      token,
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
}

// SetPublicKey sets Monobank public key for signature verification
func (c *MonobankClient) SetPublicKey(pemKey string) error {
	block, _ := pem.Decode([]byte(pemKey))
	if block == nil {
		return fmt.Errorf("failed to parse PEM block")
	}

	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return err
	}

	ecdsaPub, ok := pub.(*ecdsa.PublicKey)
	if !ok {
		return fmt.Errorf("key is not ECDSA")
	}

	c.publicKey = ecdsaPub
	return nil
}

// Name returns provider name
func (c *MonobankClient) Name() string { return "monobank" }

// CreatePayment creates Monobank payment
func (c *MonobankClient) CreatePayment(ctx context.Context, req *PaymentRequest) (*PaymentResponse, error) {
	// Monobank uses kopiyky (1 UAH = 100 kopiyky)
	amountKopiyky := int(req.Amount * 100)

	payload := map[string]interface{}{
		"amount":     amountKopiyky,
		"ccy":        currencyCode(req.Currency),
		"merchantPaymInfo": map[string]interface{}{
			"reference":   req.OrderID,
			"destination": req.Description,
		},
	}

	if req.ReturnURL != "" {
		payload["redirectUrl"] = req.ReturnURL
	}
	if req.CallbackURL != "" {
		payload["webHookUrl"] = req.CallbackURL
	}

	// Payment methods
	if req.Method != "" {
		switch req.Method {
		case MethodApplePay:
			payload["paymentType"] = "applepay"
		case MethodGooglePay:
			payload["paymentType"] = "googlepay"
		case MethodQR:
			payload["qrId"] = req.OrderID
		}
	}

	// Basket items
	if len(req.Products) > 0 {
		basketOrder := make([]map[string]interface{}, len(req.Products))
		for i, p := range req.Products {
			basketOrder[i] = map[string]interface{}{
				"name":  p.Name,
				"qty":   float64(p.Quantity),
				"sum":   int(p.Price * float64(p.Quantity) * 100),
				"icon":  "",
				"unit":  "шт.",
				"code":  fmt.Sprintf("item_%d", i),
			}
		}
		payload["merchantPaymInfo"].(map[string]interface{})["basketOrder"] = basketOrder
	}

	// Validity period
	payload["validity"] = 3600 // 1 hour

	// Save card token
	if req.Metadata != nil && req.Metadata["save_card"] == "true" {
		payload["saveCardData"] = map[string]interface{}{
			"saveCard": true,
		}
	}

	data, _ := json.Marshal(payload)

	httpReq, _ := http.NewRequestWithContext(ctx, "POST", monoAcquiringURL+"/invoice/create", bytes.NewReader(data))
	httpReq.Header.Set("X-Token", c.token)
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	if resp.StatusCode != 200 {
		var errResp map[string]interface{}
		json.Unmarshal(body, &errResp)
		return nil, fmt.Errorf("monobank error: %v", errResp["errText"])
	}

	var result map[string]interface{}
	json.Unmarshal(body, &result)

	return &PaymentResponse{
		PaymentID:  fmt.Sprintf("%v", result["invoiceId"]),
		OrderID:    req.OrderID,
		Status:     StatusPending,
		Amount:     req.Amount,
		Currency:   req.Currency,
		PaymentURL: fmt.Sprintf("%v", result["pageUrl"]),
		CreatedAt:  time.Now(),
	}, nil
}

// GetPayment gets payment status
func (c *MonobankClient) GetPayment(ctx context.Context, invoiceID string) (*PaymentResponse, error) {
	httpReq, _ := http.NewRequestWithContext(ctx, "GET", monoAcquiringURL+"/invoice/status?invoiceId="+invoiceID, nil)
	httpReq.Header.Set("X-Token", c.token)

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)

	status := c.mapStatus(fmt.Sprintf("%v", result["status"]))

	response := &PaymentResponse{
		PaymentID: invoiceID,
		Status:    status,
		CreatedAt: time.Now(),
	}

	if amount, ok := result["amount"].(float64); ok {
		response.Amount = amount / 100
	}
	if ccy, ok := result["ccy"].(float64); ok {
		response.Currency = currencyFromCode(int(ccy))
	}
	if failureReason, ok := result["failureReason"].(string); ok {
		response.ErrorMessage = failureReason
	}

	// Get reference from merchantPaymInfo
	if info, ok := result["merchantPaymInfo"].(map[string]interface{}); ok {
		if ref, ok := info["reference"].(string); ok {
			response.OrderID = ref
		}
	}

	return response, nil
}

// ProcessCallback processes Monobank webhook
func (c *MonobankClient) ProcessCallback(ctx context.Context, data []byte, signature string) (*PaymentCallback, error) {
	// Verify signature if public key is set
	if c.publicKey != nil && signature != "" {
		if !c.VerifySignature(data, signature) {
			return nil, ErrCallbackInvalid
		}
	}

	var payload map[string]interface{}
	if err := json.Unmarshal(data, &payload); err != nil {
		return nil, err
	}

	status := c.mapStatus(fmt.Sprintf("%v", payload["status"]))

	callback := &PaymentCallback{
		PaymentID: fmt.Sprintf("%v", payload["invoiceId"]),
		Status:    status,
		RawData:   payload,
	}

	if amount, ok := payload["amount"].(float64); ok {
		callback.Amount = amount / 100
	}
	if ccy, ok := payload["ccy"].(float64); ok {
		callback.Currency = currencyFromCode(int(ccy))
	}

	// Get reference
	if ref, ok := payload["reference"].(string); ok {
		callback.OrderID = ref
	}

	// Card info
	if cardMask, ok := payload["maskedPan"].(string); ok {
		callback.CardMask = cardMask
	}

	// Failure reason
	if failureReason, ok := payload["failureReason"].(string); ok {
		callback.ErrorMessage = failureReason
	}

	return callback, nil
}

// Refund creates refund (cancellation)
func (c *MonobankClient) Refund(ctx context.Context, req *RefundRequest) (*RefundResponse, error) {
	payload := map[string]interface{}{
		"invoiceId": req.PaymentID,
	}

	if req.Amount > 0 {
		payload["amount"] = int(req.Amount * 100)
	}
	if req.Reason != "" {
		payload["extRef"] = req.Reason
	}

	data, _ := json.Marshal(payload)

	httpReq, _ := http.NewRequestWithContext(ctx, "POST", monoAcquiringURL+"/invoice/cancel", bytes.NewReader(data))
	httpReq.Header.Set("X-Token", c.token)
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		var errResp map[string]interface{}
		json.Unmarshal(body, &errResp)
		return nil, fmt.Errorf("refund failed: %v", errResp["errText"])
	}

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)

	return &RefundResponse{
		RefundID:  fmt.Sprintf("%v", result["invoiceId"]),
		PaymentID: req.PaymentID,
		Status:    StatusRefunded,
		CreatedAt: time.Now(),
	}, nil
}

// VerifySignature verifies webhook signature
func (c *MonobankClient) VerifySignature(data []byte, signature string) bool {
	if c.publicKey == nil {
		return true // Skip if no public key set
	}

	sigBytes, err := base64.StdEncoding.DecodeString(signature)
	if err != nil {
		return false
	}

	hash := sha256.Sum256(data)
	return ecdsa.VerifyASN1(c.publicKey, hash[:], sigBytes)
}

// CreateQRPayment creates QR code payment
func (c *MonobankClient) CreateQRPayment(ctx context.Context, req *PaymentRequest) (*PaymentResponse, error) {
	amountKopiyky := int(req.Amount * 100)

	payload := map[string]interface{}{
		"amount": amountKopiyky,
		"ccy":    currencyCode(req.Currency),
		"merchantPaymInfo": map[string]interface{}{
			"reference":   req.OrderID,
			"destination": req.Description,
		},
	}

	if req.CallbackURL != "" {
		payload["webHookUrl"] = req.CallbackURL
	}

	data, _ := json.Marshal(payload)

	httpReq, _ := http.NewRequestWithContext(ctx, "POST", monoAcquiringURL+"/qr/details", bytes.NewReader(data))
	httpReq.Header.Set("X-Token", c.token)
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)

	return &PaymentResponse{
		PaymentID:  fmt.Sprintf("%v", result["invoiceId"]),
		OrderID:    req.OrderID,
		Status:     StatusPending,
		Amount:     req.Amount,
		Currency:   req.Currency,
		PaymentURL: fmt.Sprintf("%v", result["pageUrl"]),
		QRCode:     fmt.Sprintf("%v", result["qrLink"]),
		CreatedAt:  time.Now(),
	}, nil
}

// Finalize finalizes 2-step payment (confirm after auth)
func (c *MonobankClient) Finalize(ctx context.Context, invoiceID string, amount float64) error {
	payload := map[string]interface{}{
		"invoiceId": invoiceID,
	}
	if amount > 0 {
		payload["amount"] = int(amount * 100)
	}

	data, _ := json.Marshal(payload)

	httpReq, _ := http.NewRequestWithContext(ctx, "POST", monoAcquiringURL+"/invoice/finalize", bytes.NewReader(data))
	httpReq.Header.Set("X-Token", c.token)
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		var errResp map[string]interface{}
		json.Unmarshal(body, &errResp)
		return fmt.Errorf("finalize failed: %v", errResp["errText"])
	}

	return nil
}

// GetMerchantInfo gets merchant info
func (c *MonobankClient) GetMerchantInfo(ctx context.Context) (map[string]interface{}, error) {
	httpReq, _ := http.NewRequestWithContext(ctx, "GET", monoAcquiringURL+"/details", nil)
	httpReq.Header.Set("X-Token", c.token)

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)

	return result, nil
}

// GetStatement gets merchant statement
func (c *MonobankClient) GetStatement(ctx context.Context, from, to time.Time) ([]map[string]interface{}, error) {
	url := fmt.Sprintf("%s/statement?from=%d&to=%d",
		monoAcquiringURL,
		from.Unix(),
		to.Unix())

	httpReq, _ := http.NewRequestWithContext(ctx, "GET", url, nil)
	httpReq.Header.Set("X-Token", c.token)

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result struct {
		List []map[string]interface{} `json:"list"`
	}
	json.NewDecoder(resp.Body).Decode(&result)

	return result.List, nil
}

func (c *MonobankClient) mapStatus(status string) PaymentStatus {
	switch status {
	case "success":
		return StatusSuccess
	case "failure":
		return StatusFailed
	case "reversed":
		return StatusRefunded
	case "created", "processing", "hold":
		return StatusProcessing
	case "expired":
		return StatusExpired
	default:
		return StatusPending
	}
}

func currencyCode(cur Currency) int {
	switch cur {
	case UAH:
		return 980
	case USD:
		return 840
	case EUR:
		return 978
	default:
		return 980
	}
}

func currencyFromCode(code int) Currency {
	switch code {
	case 980:
		return UAH
	case 840:
		return USD
	case 978:
		return EUR
	default:
		return UAH
	}
}

// MonobankPersonalClient for personal account API
type MonobankPersonalClient struct {
	token      string
	httpClient *http.Client
}

// NewMonobankPersonalClient creates personal client
func NewMonobankPersonalClient(token string) *MonobankPersonalClient {
	return &MonobankPersonalClient{
		token:      token,
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
}

// GetClientInfo gets client info
func (c *MonobankPersonalClient) GetClientInfo(ctx context.Context) (map[string]interface{}, error) {
	httpReq, _ := http.NewRequestWithContext(ctx, "GET", "https://api.monobank.ua/personal/client-info", nil)
	httpReq.Header.Set("X-Token", c.token)

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)
	return result, nil
}

// GetPersonalStatement gets account statement
func (c *MonobankPersonalClient) GetPersonalStatement(ctx context.Context, account string, from time.Time) ([]map[string]interface{}, error) {
	url := fmt.Sprintf("https://api.monobank.ua/personal/statement/%s/%d", account, from.Unix())

	httpReq, _ := http.NewRequestWithContext(ctx, "GET", url, nil)
	httpReq.Header.Set("X-Token", c.token)

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result []map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)
	return result, nil
}

// CreateJar creates donation jar payment link
func (c *MonobankPersonalClient) CreateJar(ctx context.Context, jarID string, amount int) string {
	return fmt.Sprintf("https://send.monobank.ua/jar/%s?a=%s", jarID, strconv.Itoa(amount))
}

File: services/storefront/lib/payments/monobank.ts
/**
 * Monobank Acquiring Integration
 * Інтеграція з еквайрингом Monobank
 * @see https://api.monobank.ua/docs/acquiring.html
 */

import { paymentLogger } from '../logger';

export interface MonobankConfig {
    token: string;
    testMode: boolean;
}

export interface MonobankInvoiceRequest {
    amount: number; // Сума в копійках
    ccy?: number; // Код валюти ISO 4217 (980 = UAH)
    merchantPaymInfo?: {
        reference?: string; // Номер замовлення
        destination?: string; // Призначення платежу
        comment?: string;
        customerEmails?: string[];
        basketOrder?: MonobankBasketItem[];
    };
    redirectUrl?: string;
    webHookUrl?: string;
    validity?: number; // Термін дії рахунку в секундах
    paymentType?: 'debit' | 'hold';
    qrId?: string;
    code?: string; // QR-каса
    saveCardData?: {
        saveCard: boolean;
        walletId?: string;
    };
}

export interface MonobankBasketItem {
    name: string;
    qty: number;
    sum: number; // Сума в копійках
    icon?: string;
    unit?: string;
    code?: string;
    barcode?: string;
    header?: string;
    footer?: string;
    tax?: number[];
    uktzed?: string;
}

export interface MonobankInvoiceResponse {
    invoiceId: string;
    pageUrl: string;
}

export interface MonobankInvoiceStatus {
    invoiceId: string;
    status: MonobankPaymentStatus;
    failureReason?: string;
    errCode?: string;
    amount: number;
    ccy: number;
    finalAmount?: number;
    createdDate: string;
    modifiedDate: string;
    reference?: string;
    destination?: string;
    cancelList?: MonobankCancelItem[];
    walletData?: {
        cardToken: string;
        walletId: string;
        status: string;
    };
}

export type MonobankPaymentStatus =
    | 'created'
    | 'processing'
    | 'hold'
    | 'success'
    | 'failure'
    | 'reversed'
    | 'expired';

export interface MonobankCancelItem {
    status: 'processing' | 'success' | 'failure';
    amount: number;
    ccy: number;
    createdDate: string;
    modifiedDate: string;
    approvalCode?: string;
    rrn?: string;
    extRef?: string;
}

export interface MonobankWebhookPayload {
    invoiceId: string;
    status: MonobankPaymentStatus;
    failureReason?: string;
    errCode?: string;
    amount: number;
    ccy: number;
    finalAmount?: number;
    createdDate: string;
    modifiedDate: string;
    reference?: string;
}

export interface MonobankRefundRequest {
    invoiceId: string;
    amount?: number; // Якщо не вказано - повне повернення
    extRef?: string;
}

export interface MonobankMerchantInfo {
    merchantId: string;
    merchantName: string;
    edrpou?: string;
    pubKey?: {
        keyId: string;
        key: string;
    };
}

class MonobankService {
    private config: MonobankConfig;
    private apiUrl = 'https://api.monobank.ua/api/merchant';

    constructor(config?: Partial<MonobankConfig>) {
        this.config = {
            token: process.env.MONOBANK_TOKEN || '',
            testMode: process.env.NODE_ENV !== 'production',
            ...config,
        };
    }

    /**
     * Створення рахунку на оплату
     */
    async createInvoice(request: MonobankInvoiceRequest): Promise<MonobankInvoiceResponse> {
        const payload = {
            amount: request.amount,
            ccy: request.ccy || 980, // UAH
            merchantPaymInfo: request.merchantPaymInfo,
            redirectUrl: request.redirectUrl || process.env.NEXT_PUBLIC_SITE_URL + '/checkout/result',
            webHookUrl: request.webHookUrl || process.env.NEXT_PUBLIC_SITE_URL + '/api/payments/monobank/webhook',
            validity: request.validity || 3600, // 1 година
            paymentType: request.paymentType || 'debit',
            ...(request.saveCardData && { saveCardData: request.saveCardData }),
        };

        const response = await this.sendRequest('/invoice/create', 'POST', payload);
        return response as MonobankInvoiceResponse;
    }

    /**
     * Створення рахунку для замовлення
     */
    async createOrderInvoice(
        orderId: string,
        amount: number,
        items: Array<{ name: string; quantity: number; price: number }>,
        customerEmail?: string
    ): Promise<MonobankInvoiceResponse> {
        const basketOrder: MonobankBasketItem[] = items.map(item => ({
            name: item.name,
            qty: item.quantity,
            sum: Math.round(item.price * 100), // Конвертація в копійки
            unit: 'шт.',
        }));

        return this.createInvoice({
            amount: Math.round(amount * 100), // Конвертація в копійки
            merchantPaymInfo: {
                reference: orderId,
                destination: `Оплата замовлення #${orderId}`,
                basketOrder,
                customerEmails: customerEmail ? [customerEmail] : undefined,
            },
        });
    }

    /**
     * Перевірка статусу рахунку
     */
    async getInvoiceStatus(invoiceId: string): Promise<MonobankInvoiceStatus> {
        const response = await this.sendRequest('/invoice/status', 'GET', { invoiceId });
        return response as MonobankInvoiceStatus;
    }

    /**
     * Скасування/Повернення оплати
     */
    async cancelInvoice(request: MonobankRefundRequest): Promise<MonobankCancelItem> {
        const payload: Record<string, unknown> = {
            invoiceId: request.invoiceId,
        };

        if (request.amount) {
            payload.amount = request.amount;
        }

        if (request.extRef) {
            payload.extRef = request.extRef;
        }

        const response = await this.sendRequest('/invoice/cancel', 'POST', payload);
        return response as MonobankCancelItem;
    }

    /**
     * Фіналізація суми холду
     */
    async finalizeHold(invoiceId: string, amount?: number): Promise<MonobankInvoiceStatus> {
        const payload: Record<string, unknown> = {
            invoiceId,
        };

        if (amount) {
            payload.amount = amount;
        }

        const response = await this.sendRequest('/invoice/finalize', 'POST', payload);
        return response as MonobankInvoiceStatus;
    }

    /**
     * Інформація про мерчанта
     */
    async getMerchantInfo(): Promise<MonobankMerchantInfo> {
        const response = await this.sendRequest('/details', 'GET');
        return response as MonobankMerchantInfo;
    }

    /**
     * Виписка по мерчанту
     */
    async getStatement(from: number, to?: number): Promise<MonobankInvoiceStatus[]> {
        const params: Record<string, unknown> = { from };
        if (to) params.to = to;

        const response = await this.sendRequest('/statement', 'GET', params);
        return response as MonobankInvoiceStatus[];
    }

    /**
     * Публічний ключ мерчанта
     */
    async getPublicKey(): Promise<{ keyId: string; key: string }> {
        const response = await this.sendRequest('/pubkey', 'GET');
        return response as { keyId: string; key: string };
    }

    /**
     * Верифікація webhook підпису
     */
    verifyWebhook(body: string, xSign: string, pubKeyBase64: string): boolean {
        try {
            const crypto = require('crypto');
            const pubKey = `-----BEGIN PUBLIC KEY-----\n${pubKeyBase64}\n-----END PUBLIC KEY-----`;
            const verify = crypto.createVerify('SHA256');
            verify.update(body);
            return verify.verify(pubKey, xSign, 'base64');
        } catch (error) {
            paymentLogger.error('Monobank webhook verification error', error);
            return false;
        }
    }

    /**
     * Парсинг webhook payload
     */
    parseWebhook(body: string): MonobankWebhookPayload {
        return JSON.parse(body) as MonobankWebhookPayload;
    }

    /**
     * Відправка запиту до API
     */
    private async sendRequest(
        endpoint: string,
        method: 'GET' | 'POST',
        data?: Record<string, unknown>
    ): Promise<unknown> {
        let url = `${this.apiUrl}${endpoint}`;

        const options: RequestInit = {
            method,
            headers: {
                'X-Token': this.config.token,
                'Content-Type': 'application/json',
            },
        };

        if (method === 'GET' && data) {
            const params = new URLSearchParams();
            Object.entries(data).forEach(([key, value]) => {
                params.append(key, String(value));
            });
            url += `?${params.toString()}`;
        } else if (method === 'POST' && data) {
            options.body = JSON.stringify(data);
        }

        try {
            const response = await fetch(url, options);

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.errText || `Monobank API error: ${response.status}`);
            }

            return await response.json();
        } catch (error) {
            paymentLogger.error('Monobank API error', error);
            throw error;
        }
    }

    /**
     * Перевірка чи платіж успішний
     */
    isPaymentSuccessful(status: MonobankPaymentStatus): boolean {
        return status === 'success';
    }

    /**
     * Перевірка чи платіж в обробці
     */
    isPaymentPending(status: MonobankPaymentStatus): boolean {
        return ['created', 'processing', 'hold'].includes(status);
    }

    /**
     * Перевірка чи платіж невдалий
     */
    isPaymentFailed(status: MonobankPaymentStatus): boolean {
        return ['failure', 'expired'].includes(status);
    }

    /**
     * Конвертація копійок в гривні
     */
    fromKopecks(amount: number): number {
        return amount / 100;
    }

    /**
     * Конвертація гривень в копійки
     */
    toKopecks(amount: number): number {
        return Math.round(amount * 100);
    }
}

// Singleton instance
export const monobank = new MonobankService();

// React hook
export function useMonobank() {
    return monobank;
}
